File: static/cpp/valid-IP-address.cc
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <algorithm>
#include <cassert>
#include <iostream>
#include <random>
#include <string>
#include <vector>

using std::cout;
using std::default_random_engine;
using std::endl;
using std::random_device;
using std::stoi;
using std::string;
using std::uniform_int_distribution;
using std::vector;

bool IsValidPart(const string& s);

// @include
vector<string> GetValidIPAddress(const string& s) {
  vector<string> result;
  for (size_t i = 1; i < 4 && i < s.size(); ++i) {
    const string first = s.substr(0, i);
    if (IsValidPart(first)) {
      for (size_t j = 1; i + j < s.size() && j < 4; ++j) {
        const string second = s.substr(i, j);
        if (IsValidPart(second)) {
          for (size_t k = 1; i + j + k < s.size() && k < 4; ++k) {
            const string third = s.substr(i + j, k),
                         fourth = s.substr(i + j + k);
            if (IsValidPart(third) && IsValidPart(fourth)) {
              result.emplace_back(first + "." + second + "." + third + "." +
                                  fourth);
            }
          }
        }
      }
    }
  }
  return result;
}

bool IsValidPart(const string& s) {
  if (s.size() > 3) {
    return false;
  }
  // "00", "000", "01", etc. are not valid, but "0" is valid.
  if (s.front() == '0' && s.size() > 1) {
    return false;
  }
  int val = stoi(s);
  return val <= 255 && val >= 0;
}
// @exclude

int main(int argc, char** argv) {
  if (argc == 2) {
    auto result = GetValidIPAddress(argv[1]);
    for (const string& s : result) {
      cout << s << endl;
    }
  }
  auto res1 = GetValidIPAddress("255255255255");
  for (const string& s : res1) {
    cout << s << endl;
  }
  assert(res1.size() == 1);
  assert(res1.front() == "255.255.255.255");
  auto res2 = GetValidIPAddress("19216811");
  for (const string& s : res2) {
    cout << s << endl;
  }
  assert(res2.size() == 9);
  auto res3 = GetValidIPAddress("1111");
  for (const string& s : res3) {
    cout << s << endl;
  }
  assert(res3.size() == 1);
  assert(res3.front().compare("1.1.1.1") == 0);
  auto res4 = GetValidIPAddress("11000");
  for (const string& s : res4) {
    cout << s << endl;
  }
  assert(res4.size() == 2);
  sort(res4.begin(), res4.end());
  assert(res4[0].compare("1.10.0.0") == 0);
  assert(res4[1].compare("11.0.0.0") == 0);
  return 0;
}

File: static/python/valid_ip_address.py
import sys


# @include
def get_valid_ip_address(s):
    def is_valid_part(s):
        # '00', '000', '01', etc. are not valid, but '0' is valid.
        return len(s) == 1 or (s[0] != '0' and int(s) <= 255)

    result, parts = [], [None] * 4
    for i in range(1, min(4, len(s))):
        parts[0] = s[:i]
        if is_valid_part(parts[0]):
            for j in range(1, min(len(s) - i, 4)):
                parts[1] = s[i:i + j]
                if is_valid_part(parts[1]):
                    for k in range(1, min(len(s) - i - j, 4)):
                        parts[2], parts[3] = s[i + j:i + j + k], s[i + j + k:]
                        if is_valid_part(parts[2]) and is_valid_part(parts[3]):
                            result.append('.'.join(parts))
    return result
# @exclude


def main():
    if len(sys.argv) == 2:
        result = get_valid_ip_address(sys.argv[1])
        for s in result:
            print(s)
    res1 = get_valid_ip_address('255255255255')
    assert res1 == ['255.255.255.255']
    res2 = get_valid_ip_address('19216811')
    assert len(res2) == 9
    res3 = get_valid_ip_address('1111')
    assert res3 == ['1.1.1.1']
    res4 = get_valid_ip_address('11000')
    assert len(res4) == 2
    sorted(res4)
    assert res4 == ['1.10.0.0', '11.0.0.0']


if __name__ == '__main__':
    main()
